### DISCUSSION TITLE
Tagged enum in Query

### DISCUSSION THREAD

**Discussion by FintanH** [tokio-rs/axum#1945] [Q&A] [answered]
Hey :wave: 

We're working on creating a query parameter struct but one of the fields is (de)serialized as a tagged enum. So the code roughly looks like the following:
```rust
#[serde(rename_all = "camelCase")]
pub struct CobsQuery {
    pub page: Option<usize>,
    pub per_page: Option<usize>,
    pub state: Option<State>,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", tag = "status")]
pub enum State {
    #[default]
    Open,
    Draft,
    Archived,
    Merged,
}
```

We would hope to use the params like: `?page=1?perPage=10?state=open`, however, due to the `status` tag, it would expect `state={"status": "open"}`.

We can change `pub state: Option<State>` to `pub state: Option<String>` and do the parsing of the string instead, but thought we'd first check if it's possible to use the tagged enum approach somehow.

Thanks for any help you can provide! :blush:

**Comment by davidpdrsn** [comment, marked as answer]
Axum uses serde_urlencoded which I believe is limited in this regard. Try searching for issues in their repo.

**Reply by FintanH** [reply]
Thanks, that's helpful!

**Comment by FintanH** [comment]
If anyone else is looking, I came across this issue https://github.com/nox/serde_urlencoded/issues/26

Going off the helpful [comment here](https://github.com/nox/serde_urlencoded/issues/26#issuecomment-653763262), it's possible to use [`#[serde(deserialize_with = "path")]`](https://serde.rs/field-attrs.html#deserialize_with) on the field and write a different deserializer for the query params :)
